File: java/src/main/java/com/example/cote/GJM/실험데이터_복구하기.java
package com.example.cote.GJM;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class 실험데이터_복구하기 {

  private static int MAX_N = 15;

  private static int[][] overlap = new int[MAX_N][MAX_N];
  private static int[][] cache = new int[MAX_N][1 << MAX_N];
  private static List<String> words;

  private static String reconstruct(int cur, int taken) {

    if (taken == (1 << words.size()) - 1) return "";

    for (int next = 0; next < words.size(); next++) {
      if ((taken & (1 << next)) != 0) continue;
      if (restore(cur, taken) == overlap[cur][next] + restore(next, taken | (1 << next)))
        return words.get(next).substring(overlap[cur][next])
            + reconstruct(next, taken | (1 << next));
    }
    return "Error";
  }

  private static int restore(int cur, int taken) {

    if (taken == (1 << words.size()) - 1) return 0;

    int ret = cache[cur][taken];
    if (ret != -1) return ret;

    ret = 0;

    for (int next = 0; next < words.size(); next++) {
      if ((taken & (1 << next)) != 0) continue;
      ret = restore(next, taken | (1 << next)) + overlap[cur][next];
    }

    return ret;
  }

  private static int calcOverlap(String a, String b) {
    int ret = 0;

    for (int i = 0; i < a.length(); i++) {
      int len = Math.min(a.length() - i, b.length());
      if (a.substring(i, i + len).equals(b.substring(0, len))) {
        ret = Math.max(ret, len);
      }
    }

    return ret;
  }

  private static void precalc() {

    for (int i = 0; i < MAX_N; i++) {
      for (int j = 0; j < MAX_N; j++) {
        overlap[i][j] = 0;
      }
    }

    for (int i = 0; i < MAX_N; i++) {
      for (int j = 0; j < (1 << MAX_N); j++) {
        cache[i][j] = -1;
      }
    }

    for (int i = 0; i < words.size(); i++) {
      for (int j = 0; j < words.size(); j++) {
        overlap[i][j] = calcOverlap(words.get(i), words.get(j));
      }
    }
  }

  private static void solve() {

    precalc();

    int maxIdx = 0;
    int max = 0;

    for (int i = 0; i < words.size(); i++) {
      int tmp = restore(i, (1 << i));

      if (max < tmp) {
        maxIdx = i;
        max = tmp;
      }
    }

    System.out.println(words.get(maxIdx) + reconstruct(maxIdx, (1 << maxIdx)));
  }

  public static void main(String[] args) throws IOException {

    Path p = Paths.get(System.getProperty("user.dir") + "/data/실험데이터_복구하기.txt");
    BufferedReader reader = Files.newBufferedReader(p);

    int test_cases = Integer.parseInt(reader.readLine());

    while (test_cases > 0) {
      int n = Integer.parseInt(reader.readLine());
      words = new LinkedList<>();

      for (int i = 0; i < n; i++) {
        words.add(reader.readLine());
      }

      solve();

      test_cases--;
    }
  }
}

File: c++/cote/GJM/실험데이터_복구하기.cpp
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#define MAX_N 15
#define MAX_ 987654321

using namespace std;

int overlap[MAX_N][MAX_N];
int cache[MAX_N][1 << MAX_N];

int getOverlap(const string& a, const string& b) {
  int count = 0;

  for (int i = 0; i < a.size(); i++) {
    int len = min(a.size() - i, b.size());
    if (a.substr(i, len) == b.substr(0, len)) {
      count = max(count, len);
    }
  }

  return count;
}

void preCalc(const vector<string>& words) {
  for (int i = 0; i < words.size(); i++) {
    for (int j = 0; j < words.size(); j++) {
      overlap[i][j] = getOverlap(words[i], words[j]);
      // overlap[i][j] = max(getOverlap(words[i], words[j]),getOverlap(words[j],
      // words[i]));
    }
  }
}

int restore(int cur, int used, int k) {
  if (used == (1 << k) - 1) return 0;

  int ret = cache[cur][used];
  if (ret != -1) return ret;
  ret = 0;

  for (int i = 0; i < k; i++) {
    if ((used & (1 << i))) continue;

    int cand = overlap[cur][i] + restore(i, used | (1 << i), k);
    ret = max(ret, cand);
  }

  return ret;
}

string reconstruct(int cur, int used, const vector<string>& words) {
  if (used == (1 << words.size()) - 1) return "";

  for (int i = 0; i < words.size(); i++) {
    if (used & (i << i)) continue;

    int ifUsed = restore(i, used + (1 << i), words.size()) + overlap[cur][i];

    if (restore(cur, used, words.size()) == ifUsed) {
      return (words[i].substr(overlap[cur][i]) +
              reconstruct(i, used + (1 << i), words));
    }
  }
  return "Error";
}

string solve(const vector<string>& words) {
  memset(overlap, 0, sizeof(overlap));
  memset(cache, -1, sizeof(cache));
  preCalc(words);
  ` int m = -1;
  int cur_max = 0;

  for (int i = 0; i < words.size(); i++) {
    int tmp = restore(i, 1 << i, words.size());
    if (m < tmp) {
      cur_max = i;
      m = tmp;
    }
  }

  cout << words[cur_max] << reconstruct(cur_max, 1 << cur_max, words) << endl;

  return "";
}

int main(void) {
  fstream fd("data/실험데이터_복구하기.txt");
  string buf;

  getline(fd, buf);
  int test_cases = stoi(buf);

  while (test_cases > 0) {
    getline(fd, buf);

    vector<string> words;
    int nWord = stoi(buf);

    for (int i = 0; i < nWord; i++) {
      getline(fd, buf);
      words.push_back(buf);
    }

    cout << solve(words) << endl;

    test_cases--;
  }

  return 0;
}